File: google-drive/node/src/googleDriveClient.js
import fs from 'node:fs';
import path from 'node:path';
import { google } from 'googleapis';

const RETRYABLE_CODES = new Set([429, 500, 502, 503, 504]);

export class GoogleDriveClient {
  constructor({
    credentialsPath = process.env.GOOGLE_APPLICATION_CREDENTIALS,
    rootFolderId = process.env.GOOGLE_DRIVE_ROOT_FOLDER_ID,
    supportsAllDrives = ['true', '1'].includes(process.env.GOOGLE_DRIVE_SUPPORTS_ALL_DRIVES),
    maxRetries = 3,
    retryDelayMs = 500
  } = {}) {
    if (!credentialsPath) {
      throw new Error('Missing GOOGLE_APPLICATION_CREDENTIALS path.');
    }

    this.rootFolderId = rootFolderId || null;
    this.supportsAllDrives = Boolean(supportsAllDrives);
    this.maxRetries = maxRetries;
    this.retryDelayMs = retryDelayMs;

    const auth = new google.auth.GoogleAuth({
      keyFile: credentialsPath,
      scopes: ['https://www.googleapis.com/auth/drive']
    });

    this.drive = google.drive({ version: 'v3', auth });
  }

  async uploadFile(localPath, mimeType, parentFolderId = null) {
    if (!fs.existsSync(localPath)) {
      throw new Error(`File not found: ${localPath}`);
    }

    const parents = this.buildParents(parentFolderId);
    const requestBody = { name: path.basename(localPath), parents };

    return this.withRetry(async () => {
      const response = await this.drive.files.create({
        requestBody,
        media: { mimeType, body: fs.createReadStream(localPath) },
        fields: 'id,name,parents,webViewLink',
        supportsAllDrives: this.supportsAllDrives
      });

      return response.data;
    }, 'uploadFile');
  }

  async listFiles(query = null, pageSize = 20) {
    const parts = [];
    if (this.rootFolderId) {
      parts.push(`'${this.rootFolderId}' in parents`);
    }
    if (query) {
      parts.push(query);
    }
    const q = parts.length ? parts.join(' and ') : undefined;

    return this.withRetry(async () => {
      const response = await this.drive.files.list({
        q,
        pageSize,
        fields: 'files(id,name,mimeType,modifiedTime,parents)',
        supportsAllDrives: this.supportsAllDrives,
        includeItemsFromAllDrives: this.supportsAllDrives
      });

      return response.data.files || [];
    }, 'listFiles');
  }

  async getFileMeta(fileId) {
    return this.withRetry(async () => {
      const response = await this.drive.files.get({
        fileId,
        fields: 'id,name,mimeType,size,modifiedTime,parents,webViewLink',
        supportsAllDrives: this.supportsAllDrives
      });

      return response.data;
    }, 'getFileMeta');
  }

  async downloadFile(fileId, destPath) {
    return this.withRetry(async () => {
      const response = await this.drive.files.get(
        { fileId, alt: 'media', supportsAllDrives: this.supportsAllDrives },
        { responseType: 'stream' }
      );

      await new Promise((resolve, reject) => {
        const writeStream = fs.createWriteStream(destPath);
        response.data.pipe(writeStream);
        response.data.on('error', reject);
        writeStream.on('finish', resolve);
      });
    }, 'downloadFile');
  }

  async updateFile(fileId, newLocalPath = null, newName = null) {
    if (newLocalPath && !fs.existsSync(newLocalPath)) {
      throw new Error(`File not found: ${newLocalPath}`);
    }

    return this.withRetry(async () => {
      const requestBody = {};
      if (newName) {
        requestBody.name = newName;
      }

      const params = {
        fileId,
        requestBody,
        fields: 'id,name,mimeType,modifiedTime',
        supportsAllDrives: this.supportsAllDrives
      };

      if (newLocalPath) {
        params.media = {
          mimeType: 'application/octet-stream',
          body: fs.createReadStream(newLocalPath)
        };
      }

      const response = await this.drive.files.update(params);
      return response.data;
    }, 'updateFile');
  }

  async deleteFile(fileId) {
    return this.withRetry(async () => {
      await this.drive.files.delete({ fileId, supportsAllDrives: this.supportsAllDrives });
    }, 'deleteFile');
  }

  async createFolder(name, parentFolderId = null) {
    const parents = this.buildParents(parentFolderId);

    return this.withRetry(async () => {
      const response = await this.drive.files.create({
        requestBody: {
          name,
          parents,
          mimeType: 'application/vnd.google-apps.folder'
        },
        fields: 'id,name,parents',
        supportsAllDrives: this.supportsAllDrives
      });

      return response.data;
    }, 'createFolder');
  }

  buildParents(parentFolderId) {
    const parent = parentFolderId || this.rootFolderId;
    return parent ? [parent] : undefined;
  }

  async withRetry(fn, context) {
    let attempt = 0;
    let lastError;

    while (attempt < this.maxRetries) {
      try {
        return await fn();
      } catch (error) {
        lastError = error;
        const status = error?.code || error?.response?.status;
        if (!RETRYABLE_CODES.has(status)) {
          throw new Error(`Drive API error (${context}): ${error.message}`);
        }
      }

      attempt += 1;
      await new Promise((resolve) => setTimeout(resolve, this.retryDelayMs));
    }

    throw new Error(`Drive API failed after ${this.maxRetries} attempts (${context}). ${lastError?.message || ''}`);
  }
}

File: google-drive/csharp/JamiApis.GoogleDrive/GoogleDriveClient.cs
using Google.Apis.Auth.OAuth2;
using Google.Apis.Drive.v3;
using Google.Apis.Drive.v3.Data;
using Google.Apis.Services;
using Google.Apis.Util.Store;
using Google.Apis.Upload;

namespace JamiApis.GoogleDrive;

public sealed class GoogleDriveClient
{
    private static readonly HashSet<int> RetryableCodes = new() { 429, 500, 502, 503, 504 };
    private readonly DriveService _service;
    public DriveConfig Config { get; }

    public GoogleDriveClient(
        string? credentialsPath = null,
        string? rootFolderId = null,
        bool? supportsAllDrives = null,
        int maxRetries = 3,
        int retryDelayMs = 500)
    {
        credentialsPath ??= Environment.GetEnvironmentVariable("GOOGLE_APPLICATION_CREDENTIALS");
        if (string.IsNullOrWhiteSpace(credentialsPath))
        {
            throw new ArgumentException("Missing GOOGLE_APPLICATION_CREDENTIALS path.");
        }

        var supportsEnv = Environment.GetEnvironmentVariable("GOOGLE_DRIVE_SUPPORTS_ALL_DRIVES");
        var supportsFlag = supportsAllDrives ?? (supportsEnv == "true" || supportsEnv == "1");

        Config = new DriveConfig(
            credentialsPath,
            rootFolderId ?? Environment.GetEnvironmentVariable("GOOGLE_DRIVE_ROOT_FOLDER_ID"),
            supportsFlag,
            maxRetries,
            retryDelayMs);

        var credential = GoogleCredential.FromFile(Config.CredentialsPath)
            .CreateScoped(DriveService.Scope.Drive);

        _service = new DriveService(new BaseClientService.Initializer
        {
            HttpClientInitializer = credential,
            ApplicationName = "JamiApis Google Drive C#"
        });
    }

    public async Task<File> UploadFileAsync(string localPath, string mimeType, string? parentFolderId = null)
    {
        if (!System.IO.File.Exists(localPath))
        {
            throw new ArgumentException($"File not found: {localPath}");
        }

        var fileMetadata = new File
        {
            Name = Path.GetFileName(localPath),
            Parents = BuildParents(parentFolderId)
        };

        return await WithRetryAsync(async () =>
        {
            await using var stream = new FileStream(localPath, FileMode.Open, FileAccess.Read);
            var request = _service.Files.Create(fileMetadata, stream, mimeType);
            request.SupportsAllDrives = Config.SupportsAllDrives;
            request.Fields = "id,name,parents,webViewLink";
            var result = await request.UploadAsync();
            if (result.Status != UploadStatus.Completed)
            {
                throw new IOException($"Upload failed: {result.Exception?.Message}");
            }

            return request.ResponseBody;
        }, "uploadFile");
    }

    public async Task<IList<File>> ListFilesAsync(string? query = null, int pageSize = 20)
    {
        var qParts = new List<string>();
        if (!string.IsNullOrWhiteSpace(Config.RootFolderId))
        {
            qParts.Add($"'{Config.RootFolderId}' in parents");
        }
        if (!string.IsNullOrWhiteSpace(query))
        {
            qParts.Add(query);
        }

        return await WithRetryAsync(async () =>
        {
            var request = _service.Files.List();
            request.Q = qParts.Count > 0 ? string.Join(" and ", qParts) : null;
            request.PageSize = pageSize;
            request.Fields = "files(id,name,mimeType,modifiedTime,parents)";
            request.SupportsAllDrives = Config.SupportsAllDrives;
            request.IncludeItemsFromAllDrives = Config.SupportsAllDrives;
            var response = await request.ExecuteAsync();
            return response.Files ?? new List<File>();
        }, "listFiles");
    }

    public async Task<File> GetFileMetaAsync(string fileId)
    {
        return await WithRetryAsync(async () =>
        {
            var request = _service.Files.Get(fileId);
            request.Fields = "id,name,mimeType,size,modifiedTime,parents,webViewLink";
            request.SupportsAllDrives = Config.SupportsAllDrives;
            return await request.ExecuteAsync();
        }, "getFileMeta");
    }

    public async Task DownloadFileAsync(string fileId, string destPath)
    {
        await WithRetryAsync(async () =>
        {
            var request = _service.Files.Get(fileId);
            request.SupportsAllDrives = Config.SupportsAllDrives;
            await using var stream = new FileStream(destPath, FileMode.Create, FileAccess.Write);
            await request.DownloadAsync(stream);
            return true;
        }, "downloadFile");
    }

    public async Task<File> UpdateFileAsync(string fileId, string? newLocalPath = null, string? newName = null)
    {
        if (!string.IsNullOrWhiteSpace(newLocalPath) && !System.IO.File.Exists(newLocalPath))
        {
            throw new ArgumentException($"File not found: {newLocalPath}");
        }

        var metadata = new File();
        if (!string.IsNullOrWhiteSpace(newName))
        {
            metadata.Name = newName;
        }

        return await WithRetryAsync(async () =>
        {
            if (!string.IsNullOrWhiteSpace(newLocalPath))
            {
                await using var stream = new FileStream(newLocalPath, FileMode.Open, FileAccess.Read);
                var request = _service.Files.Update(metadata, fileId, stream, "application/octet-stream");
                request.SupportsAllDrives = Config.SupportsAllDrives;
                request.Fields = "id,name,mimeType,modifiedTime";
                var result = await request.UploadAsync();
                if (result.Status != UploadStatus.Completed)
                {
                    throw new IOException($"Update failed: {result.Exception?.Message}");
                }

                return request.ResponseBody;
            }

            var metadataRequest = _service.Files.Update(metadata, fileId);
            metadataRequest.SupportsAllDrives = Config.SupportsAllDrives;
            metadataRequest.Fields = "id,name,mimeType,modifiedTime";
            return await metadataRequest.ExecuteAsync();
        }, "updateFile");
    }

    public async Task DeleteFileAsync(string fileId)
    {
        await WithRetryAsync(async () =>
        {
            var request = _service.Files.Delete(fileId);
            request.SupportsAllDrives = Config.SupportsAllDrives;
            await request.ExecuteAsync();
            return true;
        }, "deleteFile");
    }

    public async Task<File> CreateFolderAsync(string name, string? parentFolderId = null)
    {
        var metadata = new File
        {
            Name = name,
            MimeType = "application/vnd.google-apps.folder",
            Parents = BuildParents(parentFolderId)
        };

        return await WithRetryAsync(async () =>
        {
            var request = _service.Files.Create(metadata);
            request.SupportsAllDrives = Config.SupportsAllDrives;
            request.Fields = "id,name,parents";
            return await request.ExecuteAsync();
        }, "createFolder");
    }

    private IList<string>? BuildParents(string? parentFolderId)
    {
        var parent = parentFolderId ?? Config.RootFolderId;
        return string.IsNullOrWhiteSpace(parent) ? null : new List<string> { parent };
    }

    private async Task<T> WithRetryAsync<T>(Func<Task<T>> fn, string context)
    {
        Exception? lastError = null;
        for (var attempt = 0; attempt < Config.MaxRetries; attempt++)
        {
            try
            {
                return await fn();
            }
            catch (Google.GoogleApiException apiException)
            {
                lastError = apiException;
                if (!RetryableCodes.Contains(apiException.HttpStatusCode.HasValue ? (int)apiException.HttpStatusCode.Value : 0))
                {
                    throw new InvalidOperationException($"Drive API error ({context}): {apiException.Message}", apiException);
                }
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException($"Unexpected error ({context}): {ex.Message}", ex);
            }

            await Task.Delay(Config.RetryDelayMs);
        }

        throw new InvalidOperationException($"Drive API failed after {Config.MaxRetries} attempts ({context}).", lastError);
    }
}
